### PR TITLE
Update group memberships during refresh for upstream OIDC providers

### LINKED ISSUE #863: Upstream OIDC IDP group changes should be reflected downstream
Purpose: Changes in Upstream IDP  to the user's group information should be reflected downstream.
Given that I am running a Supervisor with an upstream OIDC IDP configured,
And given that I have already used the Pinniped CLI to get tokens from the Supervisor,
And given that upstream my user's group membership changes
When I use the Pinniped CLI to refresh my downstream session
Then my group membership changes and is reflected in the user information and in the session downstream 

**Is your feature request related to a problem? Please describe.**
Changes to upstream group membership needs to be reflected downstream before refresh token expires 

**Describe the solution you'd like**
Upstream group membership changes are reflected in User Information downstream.  

**Describe alternatives you've considered**
 
**Are you considering submitting a PR for this feature?**

- **How will this project improvement be tested?**
- **How does this change the current architecture?**
- **How will this change be backwards compatible?**
- **How will this feature be documented?**

**Additional context**
Usecases: Adding Users to Groups other than existing group after initial login
                  Remove user from group while user is logged in. User is able to continue with the session but any new access request after the change is denied and user has to re-login.

### PR DESCRIPTION
**Pull request by cfryanr**
Update the user's group memberships when possible. Note that we won't always have enough information to be able to update it (see code comments). Fixes #863.

**Release note**:

This is a small part of the overall upstream refresh epic.

```release-note
The Pinniped Supervisor will update the user's group memberships when they change in the upstream OIDC provider, when possible. This happens during the downstream refresh, so it will take a few minutes when a user is added/removed to/from a group in the OIDC provider before the change appears on their identity in the Kubernetes cluster.
```

### CODE REVIEW

File: internal/oidc/token/token_handler_test.go [hunk 1]
```diff
@@ -1199,12 +1209,100 @@ func TestRefreshGrant(t *testing.T) {
...
 					wantRequestedScopes:               []string{"openid", "offline_access"},
 					wantGrantedScopes:                 []string{"openid", "offline_access"},
+					wantGroups:                        goodGroups,
 					wantUpstreamRefreshCall:           happyOIDCUpstreamRefreshCall(),
 					wantUpstreamOIDCValidateTokenCall: happyUpstreamValidateTokenCall(refreshedUpstreamTokensWithRefreshTokenWithoutIDToken(), false),
 					wantCustomSessionDataStored:       upstreamOIDCCustomSessionDataWithNewRefreshToken(oidcUpstreamRefreshedRefreshToken),
 				},
 			},
 		},
+		{
```
Comment (margocrawf) [inline internal/oidc/token/token_handler_test.go:1219]: Maybe add a test for when the groups claim exists in the upstream refresh request but has an empty list as the value.

Comment (cfryanr) [reply internal/oidc/token/token_handler_test.go:1226]: Fixed!

Thanks for pointing that out because it also caused me to realize that any errors when parsing the upstream refresh groups claim would be returned in the wrong format and with the wrong status code. Added tests and new production code for that too.

**Review by margocrawf — APPROVED** [review]
